src/test/ast2sql_operation_based_conversion_test.ml

```diff
@@ -0,0 +1,139 @@
+                  Not (Pred ("eed", [ NamedVar "V1"; NamedVar "D" ]));
+                ]
+              in
+              let rule3 =
+                (* "+ed(E, D) :- ed(V1, D), E = 'Joe', D = 'A', V1 != 'Joe', ¬ed(E, D), ¬eed(V1, D)." *)
+                Deltainsert ("ed", [ NamedVar "E"; NamedVar "D" ]), [
+                  Rel (Pred ("ed", [ NamedVar "V1"; NamedVar "D" ]));
+                  Equat (Equation ("=", Var (NamedVar "E"), Const (String "'Joe'")));
+                  Equat (Equation ("=", Var (NamedVar "D"), Const (String "'A'")));
+                  Equat (Equation ("<>", Var (NamedVar "V1"), Const (String "'Joe'")));
+                  Not (Pred ("ed", [ NamedVar "E"; NamedVar "D" ]));
+                  Not (Pred ("eed", [ NamedVar "V1"; NamedVar "D" ]));
+                ]
+              in
+              [ rule3; rule2; rule1 ] (* `expr` holds its rules in the reversed order *)
+            end;
+            facts = [];
+            query = None;
+            sources = [
+              ("ed", [ ("emp_name", Sstring); ("dept_name", Sstring) ]);
+              ("eed", [ ("emp_name", Sstring); ("dept_name", Sstring) ]);
+            ];
+            view = None;
+            constraints = [];
+            primary_keys = [];
+          };
+        expected =
+          let query1 =
+            String.concat " " [
+              "SELECT ed_0.emp_name AS emp_name, 'A' AS dept_name FROM ed AS ed_0 WHERE";
+              "ed_0.dept_name = 'A' AND ed_0.emp_name <> 'Joe' AND";
+              "NOT EXISTS ( SELECT * FROM eed AS t WHERE t.emp_name = ed_0.emp_name AND t.dept_name = 'A' )";
+            ]
+          in
+          let query2 =
+            String.concat " " [
+              "SELECT 'Joe' AS emp_name, 'A' AS dept_name FROM ed AS ed_0, eed AS eed_1 WHERE";
+              "ed_0.dept_name = 'A' AND eed_1.dept_name = 'A' AND eed_1.emp_name = 'Joe' AND ed_0.emp_name <> 'Joe' AND";
+              "NOT EXISTS ( SELECT * FROM eed AS t WHERE t.emp_name = ed_0.emp_name AND t.dept_name = 'A' )";
+            ]
+          in
+          let query3 =
+            String.concat " " [
+              "SELECT 'Joe' AS emp_name, 'A' AS dept_name FROM ed AS ed_0 WHERE";
+              "ed_0.dept_name = 'A' AND ed_0.emp_name <> 'Joe' AND";
+              "NOT EXISTS ( SELECT * FROM ed AS t WHERE t.emp_name = 'Joe' AND t.dept_name = 'A' ) AND";
+              "NOT EXISTS ( SELECT * FROM eed AS t WHERE t.emp_name = ed_0.emp_name AND t.dept_name = 'A' )";
+            ]
+          in
+          String.concat " " [
+            Printf.sprintf "CREATE TEMPORARY TABLE temp0 AS %s;" query1;
+            Printf.sprintf "CREATE TEMPORARY TABLE temp1 AS %s;" query2;
+            Printf.sprintf "CREATE TEMPORARY TABLE temp2 AS %s;" query3;
+            "INSERT INTO temp0 SELECT * FROM temp0 AS inst;";
+            "DELETE FROM temp1 WHERE EXISTS ( SELECT * FROM temp1 AS inst );";
+            "INSERT INTO temp2 SELECT * FROM temp2 AS inst;";
```

Comment:
* What is `AS inst`? Is it necessary?
* `rule1` means add to `ed`, so the corresponding `query1` must be `INSERT INTO ed`, same for `temp2` and `DELETE` statement.
* The `DELETE` statement didn't work in the original spec, if it existed, it would all disappear, what I want to do is something like `eed - temp1`, can SQL do something with the `DELETE WHERE` clause (instead of `EXISTS`)?

Reply:
Thank you all for pointing out!

Firstly [c1f9657](https://github.com/dangtv/BIRDS/pull/72/commits/c1f96579859be8124d0a32134f0b1a19dc6aae8e) remedies the second problem.

Reply:
[205e15a](https://github.com/dangtv/BIRDS/pull/72/commits/205e15ad61cd63c18eeff2d35c550ce8542ac9e3) remedies the first issue.

Reply:
06c3b1a fixes the third issue, i.e., makes the program output `DELETE` queries appropriate in operational terms (which do not seem so performant though).